### PR TITLE
Fix various problems related to equals/hashCode, or using inappropriate values (wrong type even) in collections operations

### PR DESCRIPTION
**Pull request by srowen**
(Let me know if this type of change is of interest -- there's some more modest improvements that could come from some more static analysis.)

Here I found some potential equals/hashCode problems, and probable bugs based on wrong types used in collections. I'll annotate some of the changes below.

### CODE REVIEW

File: deeplearning4j-core/src/main/java/org/deeplearning4j/util/StringCluster.java [hunk 1]
```diff
@@ -78,7 +78,7 @@ public void sort() {
         public int compare(Map<String,Integer> o1, Map<String,Integer> o2) {
             int s1 = o1.size();
             int s2 = o2.size();
-            if (o1 == o2) {
+            if (s1 == s2) {
```
Comment (srowen) [inline deeplearning4j-core/src/main/java/org/deeplearning4j/util/StringCluster.java]: I assume this was the intent, to compare values only if sizes are the same

File: deeplearning4j-core/src/main/java/org/deeplearning4j/util/StringCluster.java [hunk 2]
```diff
@@ -87,9 +87,13 @@ public int compare(Map<String,Integer> o1, Map<String,Integer> o2) {
                 for (int i : o2.values()) {
                     total2 += i;
                 }
-                return total2 - total1;
-            } else {
-                return s2 - s1;
+                if (total2 < total1) return -1;
```
Comment (srowen) [inline deeplearning4j-core/src/main/java/org/deeplearning4j/util/StringCluster.java]: I think overflow is a real potential concern here, enough that subtraction is probably a potential bug

File: deeplearning4j-core/src/main/java/org/deeplearning4j/util/StringGrid.java [hunk 3]
```diff
@@ -404,7 +404,7 @@ private void modifyRows(Set<Integer> alreadyDeDupped,Integer column,List<Integer
     private String maximalValue(Map<String,Integer> map) {
         Counter<String> counter = new Counter<>();
         for (Map.Entry<String, Integer> entry : map.entrySet()) {
-            counter.incrementCount(entry.getKey(), map.get(entry.getValue()));
+            counter.incrementCount(entry.getKey(), entry.getValue());
```
Comment (srowen) [inline deeplearning4j-core/src/main/java/org/deeplearning4j/util/StringGrid.java]: Clearly intended to access the integer value, which can't also be a key in the map (which are strings)

File: deeplearning4j-scaleout/deeplearning4j-nlp/src/main/java/org/deeplearning4j/models/word2vec/VocabWord.java [hunk 4]
```diff
@@ -91,7 +91,7 @@ public void setWord(String word) {
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
-        if (o == null) return false;
+        if (!(o instanceof VocabWord)) return false;
```
Comment (srowen) [inline deeplearning4j-scaleout/deeplearning4j-nlp/src/main/java/org/deeplearning4j/models/word2vec/VocabWord.java]: This covers the case of null too BTW

File: deeplearning4j-scaleout/deeplearning4j-nlp/src/main/java/org/deeplearning4j/models/word2vec/wordstore/inmemory/AbstractCache.java [hunk 5]
```diff
@@ -386,8 +386,9 @@ public void updateWordsOccurencies() {
     @Override
     public void removeElement(String label) {
         if (vocabulary.containsKey(label)) {
-            totalWordCount.getAndAdd((long) vocabulary.get(label).getElementFrequency() * -1);
-            idxMap.remove(label);
+            SequenceElement element = vocabulary.get(label);
+            totalWordCount.getAndAdd((long) element.getElementFrequency() * -1);
+            idxMap.remove(element.getIndex());
```
Comment (srowen) [inline deeplearning4j-scaleout/deeplearning4j-nlp/src/main/java/org/deeplearning4j/models/word2vec/wordstore/inmemory/AbstractCache.java]: Same type of problem; this map isn't indexed by labels (strings) but by indices

File: deeplearning4j-scaleout/deeplearning4j-scaleout-akka/src/main/java/org/deeplearning4j/scaleout/statetracker/hazelcast/BaseHazelCastStateTracker.java [hunk 6]
```diff
@@ -844,7 +844,7 @@ public Job jobFor(String id) {
 
     private boolean isCurrentlyJob(String id) {
         for(Job j : jobs)
-            if(j.equals(id))
+            if(j.workerId().equals(id))
```
Comment (srowen) [inline deeplearning4j-scaleout/deeplearning4j-scaleout-akka/src/main/java/org/deeplearning4j/scaleout/statetracker/hazelcast/BaseHazelCastStateTracker.java]: Same, id is a String so must compare, I assume, to the string workerId

File: deeplearning4j-scaleout/deeplearning4j-scaleout-akka/src/main/java/org/deeplearning4j/scaleout/statetracker/workretriever/LocalWorkRetriever.java [hunk 7]
```diff
@@ -115,7 +115,7 @@ public Job load(String worker) {
         if(f == null || !f.exists())
             return null;
         Job d = SerializationUtils.readObject(f);
-        workerData.remove(f);
+        workerData.remove(worker);
```
Comment (srowen) [inline deeplearning4j-scaleout/deeplearning4j-scaleout-akka/src/main/java/org/deeplearning4j/scaleout/statetracker/workretriever/LocalWorkRetriever.java]: Same, workerData has worker Strings, not Files

**Comment by agibsonccc** [comment]
Yeah we can always use this kind of stuff. Thanks!

**Pull request merged**